packages/tpetra/core/src/Tpetra_CrsGraph_def.hpp

```diff
@@ -5779,137 +5781,167 @@ namespace Tpetra {
-      "failed to insert one or more indices in to padding map");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (padding.failed_insert(), std::runtime_error,
+       "failed to insert one or more indices in to padding map");
+
     return padding;
   }
 
   template <class LocalOrdinal, class GlobalOrdinal, class Node>
-  Kokkos::UnorderedMap<LocalOrdinal, size_t, typename Node::device_type>
+  void
   CrsGraph<LocalOrdinal, GlobalOrdinal, Node>::
-  computeCrsPaddingNew (const RowGraph<LocalOrdinal,GlobalOrdinal,Node>& source,
-                        const size_t numSameIDs,
-                        const Kokkos::DualView<const local_ordinal_type*,
-                          buffer_device_type>& permuteToLIDs,
-                        const Kokkos::DualView<const local_ordinal_type*,
-                          buffer_device_type>& permuteFromLIDs)
+  computeCrsPaddingForSameIDs (Kokkos::UnorderedMap<LocalOrdinal, size_t, typename Node::device_type>& padding,
+                               const RowGraph<LocalOrdinal,GlobalOrdinal,Node>& source,
+                               const size_t numSameIDs,
+                               const bool padAll) const
   {
     using LO = LocalOrdinal;
     using GO = GlobalOrdinal;
     using execution_space = typename device_type::execution_space;
-    const char tfecfFuncName[] = "computeCrsPaddingNew: ";
+    const char tfecfFuncName[] = "computeCrsPaddingForSameIds: ";
 
     execution_space::fence ();
 
-    // Resize row pointers and indices to accommodate incoming data
+    using insert_result =
+      typename Kokkos::UnorderedMap<LocalOrdinal, size_t, typename Node::device_type>::insert_result;
+
+    // Compute extra capacity needed to accommodate incoming data
+    Teuchos::Array<GO> src_row_inds;
+    Teuchos::Array<GO> tgt_row_inds;
     const map_type& src_row_map = * (source.getRowMap ());
-    using padding_type = Kokkos::UnorderedMap<LO, size_t, device_type>;
-    padding_type padding (numSameIDs + permuteFromLIDs.extent (0));
-    for (LO tgtid = 0; tgtid < static_cast<LO> (numSameIDs); ++tgtid) {
-      const GO srcgid = src_row_map.getGlobalElement (tgtid);
-      auto how_much_padding = source.getNumEntriesInGlobalRow (srcgid);
-      auto result = padding.insert (tgtid, how_much_padding);
-      // FIXME (mfh 09 Apr 2019) Kokkos::UnorderedMap is allowed to
-      // fail even if the user did nothing wrong.  We should actually
-      // have a retry option.  I just copied this code over from
-      // computeCrsPadding.
+    for (LO tgt_lid = 0; tgt_lid < static_cast<LO> (numSameIDs); ++tgt_lid) {
+      const GO src_gid = src_row_map.getGlobalElement(tgt_lid);
+      auto num_src_entries = source.getNumEntriesInGlobalRow(src_gid);
+
+      if (num_src_entries == 0)
+        continue;
+
+      insert_result result;
+      const GO tgt_gid = rowMap_->getGlobalElement(tgt_lid);
+      if (padAll) {
+        result = padding.insert(tgt_lid, num_src_entries);
```

Comment:
Right now, we just want things to compile and pass tests.  All the following comments are for the future.

I think you could use two 1-D `Kokkos::View`s to store padding information.  There are actually two sets of data over which to compute padding: sames, and permutes.  There's no reason why you need to store both in a single data structure, as long as the client can map from local row index to required padding.  The first data structure, for sames, could just be a 1-D `Kokkos::View`, since it's just indexed by `0 .. numSameIDs - 1`.  That would let you use a much smaller hash table for the second data structure.  Alternately, you could index the "permutes" padding data structure by the k in `permuteFromLIDs[k]`.  That would let you use a 1-D `Kokkos::View` as well.

Reply:
yep, when this is done, I have some ideas along the lines you mention above to get rid of `UnorderedMap` in favor of a `Kokkos::View`